firedrake/parameters.py

```diff
@@ -37,7 +37,10 @@ def rename(self, name):
     def __getstate__(self):
         # Remove non-picklable update function slot
         d = self.__dict__.copy()
-        del d["_update_function"]
+        try:
+            del d["_update_function"]
+        except KeyError:
+            pass
```

Comment:
why is this needed? I never understood this piece of code.

Reply:
This allows test grouping based on runtimes, see [the docs](https://jerry-git.github.io/pytest-split/) for details of the specific algorithm

Reply:
I was hitting an error where `_update_function` wasn't present. Note that this case is handled by `__setitem__` so missing `_update_function` seems to have been encountered before!

I also don't fully understand this code, but based on previous conversations about this, I think parameters is ripe for removal!